File: MyMinimalApi/Transfer/Domain/Commands.cs
using OneOf;
using Lib.Types;

namespace Bank.Transfer.Domain;

using TransferRecipientEvent = OneOf<
   RegisteredInternalTransferRecipient,
   RegisteredDomesticTransferRecipient,
   RegisteredInternationalTransferRecipient
>;

public record TransferCmd(
   Guid EntityId,
   string Beneficiary,
   string Iban,
   string Bic,

   DateTime Date,
   decimal Amount,
   string Reference
)
: Command(EntityId)
{
   public DebitedTransfer ToEvent() => new(
      EntityId: EntityId,
      Date: Date,
      Beneficiary: Beneficiary,
      Bic: Bic,
      DebitedAmount: Amount,
      Iban: Iban,
      Reference: Reference,
      Timestamp: Timestamp
   );
}

public record RegisterTransferRecipientCmd(
   Guid EntityId,
   TransferRecipient Recipient
)
: Command(EntityId)
{
   public TransferRecipientEvent ToEvent() =>
      Recipient.AccountEnvironment switch {
         RecipientAccountEnvironment.Internal => new RegisteredInternalTransferRecipient(
            EntityId: EntityId,
            Timestamp: Timestamp,
            LastName: Recipient.LastName,
            FirstName: Recipient.FirstName,
            AccountNumber: Recipient.Identification
         ),
         RecipientAccountEnvironment.Domestic => new RegisteredDomesticTransferRecipient(
            EntityId: EntityId,
            Timestamp: Timestamp,
            LastName: Recipient.LastName,
            FirstName: Recipient.FirstName,
            AccountNumber: Recipient.Identification,
            RoutingNumber: Recipient.RoutingNumber
         ),
         RecipientAccountEnvironment.International => new RegisteredInternationalTransferRecipient(
            EntityId: EntityId,
            LastName: Recipient.LastName,
            FirstName: Recipient.FirstName,
            Identification: Recipient.Identification,
            IdentificationStrategy: Recipient.IdentificationStrategy,
            Currency: Recipient.Currency,
            Timestamp: Timestamp
         )
      };
}

File: FSharpWithAkka/Transfer/Domain/Commands.fs
namespace Bank.Transfer.Domain

open System

open Lib.Types

type AckReceipt = string

type RecipientAccountEnvironment =
   | Internal = 0
   | Domestic = 1
   | International = 2

type RecipientAccountIdentificationStrategy =
   | AccountId = 0
   | SwiftBIC = 1
   | IBAN = 2
   | NationalID = 3

type TransferRecipient = {
   LastName: string
   FirstName: string
   Identification: string
   AccountEnvironment: RecipientAccountEnvironment
   IdentificationStrategy: RecipientAccountIdentificationStrategy
   RoutingNumber: string option
   Currency: string
}

type TransferCommand
   (
      entityId,
      correlationId,
      recipient: TransferRecipient,
      date: DateTime,
      amount: decimal,
      reference: string
   ) =
   inherit Command(entityId, correlationId)
   member x.Recipient = recipient
   member x.Date = date
   member x.Amount = amount
   member x.Reference = reference

type ApproveTransferCommand
   (
      entityId,
      correlationId,
      recipient: TransferRecipient,
      date: DateTime,
      amount: decimal,
      ackReceipt: AckReceipt
   ) =
   inherit Command(entityId, correlationId)
   member x.Recipient = recipient
   member x.Date = date
   member x.Amount = amount
   member x.AckReceipt = ackReceipt

type RejectTransferCommand
   (
      entityId,
      correlationId,
      recipient: TransferRecipient,
      date: DateTime,
      amount: decimal,
      reason: string
   ) =
   inherit Command(entityId, correlationId)
   member x.Recipient = recipient
   member x.Date = date
   member x.Amount = amount
   member x.Reason = reason

type RegisterTransferRecipientCommand
   (entityId, recipient: TransferRecipient, correlationId) =
   inherit Command(entityId, correlationId)

   member x.Recipient = {
      recipient with
         Currency =
            if isNull recipient.Currency then
               "USD"
            else
               recipient.Currency
   }
